addonscript/src/addonscript/api.py

```diff
@@ -45,29 +55,68 @@ async def add_version(context, version):
     return result
 
 
+async def check_upload(context, uuid):
+    """Check on the status of file upload identified by `uuid`
+
+    Raises:
+        SignatureError: If the upload is awaiting processing
+        FatalSignatureError: If validation errors occurred
+    """
+    url = get_api_url(context, UPLOAD_DETAIL, uuid=uuid)
+    result = await amo_get(context, url)
+    if not result["processed"]:
+        # retry
+        raise SignatureError("upload not yet processed")
+    if not result["valid"]:
+        raise FatalSignatureError(f"Automated validation produced errors: {result['validation']}")
+
+
 async def do_upload(context, locale):
     """Upload the language pack for `locale` to AMO.
 
     Returns the JSON response from AMO
     """
     locale_info = context.locales[locale]
-    langpack_id = locale_info["id"]
-    version = locale_info["version"]
-    url = get_api_url(context, UPLOAD_VERSION, id=langpack_id, version=version)
-    with open(context.locales[locale]["unsigned"], "rb") as file:
+    url = get_api_url(context, UPLOAD_CREATE)
+    with open(locale_info["unsigned"], "rb") as file:
         data = {"channel": get_channel(context.task), "upload": file}
-        try:
-            result = await amo_put(context, url, data)
-        except ClientResponseError as exc:
-            # XXX: .code is deprecated in aiohttp 3.1 in favor of .status
-            if exc.status == 409:
-                raise AMOConflictError("Addon <{}> already present on AMO with version <{}>".format(langpack_id, version))
-            # If response code is not 409 - CONFLICT, bubble the exception
-            raise exc
-        return result
-
-
-async def get_signed_addon_url(context, locale, pk):
+        return await amo_post(context, url, data)
+
+
+async def do_create_version(context, locale, upload_uuid):
+    """Create a new version for the `locale` language pack on AMO.
+
+    Returns the version's identifier
+
+    Raises:
+        AMOConflictError: If the version already exists for this addon
+        FatalSignatureError: If AMO returned a permanent error
+    """
+    locale_info = context.locales[locale]
+    langpack_id = locale_info["id"]
+    url = get_api_url(context, ADDON_CREATE_OR_EDIT, id=langpack_id)
+    data = {
+        "categories": {"firefox": ["general"]},
+        "name": {"en-US": locale_info["name"]},
+        "summary": {"en-US": locale_info["description"]},
```

Comment:
This will reset the name/summary for the addon at every upload, and change them from their current values for existing langpacks, the first time this code runs.  Overall this seems like a good thing because it gets us some consistency, and if we ever want to update the template for these fields we can do it in one place (https://searchfox.org/mozilla-central/rev/3002762e41363de8ee9ca80196d55e79651bcb6b/python/mozbuild/mozbuild/action/langpack_manifest.py#161-214)

Reply:
Yeah that seems like a good thing, along with making them all "MPL-2.0" when it first runs